src/messages/MOSDOp.h

```diff
@@ -323,7 +319,15 @@ struct ceph_osd_request_head {
       ::encode(osdmap_epoch, payload);
       ::encode(flags, payload);
       ::encode(reassert_version, payload);
-      ::encode(reqid, payload);
+      if (reqid == osd_reqid_t()) {
+	// also put the client_inc in reqid.inc, so that get_reqid() can
+	// be used before the full message is decoded.
+	osd_reqid_t fake_reqid;
+	fake_reqid.inc = inc;
+	::encode(fake_reqid, payload);
+      } else {
+	::encode(reqid, payload);
+      }
```

Comment:
d'oh. Something about this this whole setup was confusing me previously, but I agree it looks good now.